src/action-creators/flushQueue.js

```diff
@@ -0,0 +1,24 @@
+import {
+  sync,
+} from '../util'
+
+/* Returns true if the state has updates in the syncQueue */
+const hasQueue = state => state && state.syncQueue && (
+  Object.keys(state.syncQueue.thoughtIndexUpdates).length > 0 ||
+  Object.keys(state.syncQueue.contextIndexUpdates).length > 0
+)
+
+/** Syncs queued updates with the remote and clears the syncQueue. Works with action-creators/flushQueue and reducers/clearQueue */
+const syncQueue = () => (dispatch, getState) => {
+  const state = getState()
+
+  if (!hasQueue(state)) return
+
+  sync(state.syncQueue.thoughtIndexUpdates, state.syncQueue.contextIndexUpdates, { recentlyEdited: state.recentlyEdited })
+
+  dispatch({
+    type: 'clearQueue'
+  })
```

Comment:
As I understand we should clear the queue only after sync is complete. Since `sync` returns promise we can chain it to dispatch `clearQueue` only after sync is a success.

```javascript
sync()
.then(() => dispatch({ type: 'clearQueue' }))
```

Reply:
Not quite. If we don't clear the queue immediately, then `flushQueue` could get called again and the `syncQueue` could try to sync the same data. We have to clear the queue synchronously to prevent this loop.

If anything, we should store the pending queue separately so that it can be either cleared if the sync succeeds or restored if the sync fails. `Firebase` has its own internal retry mechanism, so there is no need to preserve `syncQueue` for `Firebase`. We could implement a retry/recovery mechanism for `IndexedDB`, but since that does not rely on a network connection it seems much less likely to fail, so I would consider it a lower priority.